eclair-core/src/main/scala/fr/acinq/eclair/payment/send/MultiPartPaymentLifecycle.scala

```diff
@@ -264,7 +264,7 @@ class MultiPartPaymentLifecycle(nodeParams: NodeParams, cfg: SendPaymentConfig,
           }
           paymentSent.feesPaid + localFees
       }
-      context.system.eventStream.publish(PathFindingExperimentMetrics(cfg.recipientAmount, fees, status, duration, now, isMultiPart = true, request.routeParams.experimentName, cfg.recipientNodeId))
+      context.system.eventStream.publish(PathFindingExperimentMetrics(cfg.paymentHash, cfg.recipientAmount, fees, status, duration, now, isMultiPart = true, request.routeParams.experimentName, cfg.recipientNodeId, cfg.invoice.flatMap(_.routingInfo.headOption.flatMap(_.headOption.map(_.nodeId)))))
```

Comment:
Shouldn't it be `cfg.invoice.flatMap(_.routingInfo.headOption.flatMap(_.lastOption.map(_.nodeId)))`? What exactly are you interested in? I guess it should be the node closest to the recipient, right (routing hints may contain more than one hop)?

If there are multiple routes provided in the invoice, aren't we interested in all nodes that are connected to the recipient? In which case we would want to store `routingHintNeighbors: Set[PublicKey]`, that we would compute with `cfg.invoice.flatMap(_.routingInfo.flatMap(_.lastOption.map(_.nodeId))).toSet` (to be tested, I probably wrote that too quickly and it may be buggy)?

Same in `PaymentLifecycle`.

Reply:
I am interested in the last public node on the route to the recipient. The first node of the routing hint has to be public but the others are probably private.
In the case of mobile wallets hidden behind a hub, getting the id of the mobile wallet is not very useful as we can't contact it directly, I want to know the id of the hub instead.
I am assuming here that it is extremely rare to have a private node advertising more than one hub in its routing hints, do you think it's unreasonable?

Reply:
> The first node of the routing hint has to be public but the others are probably private.

Got it, I see the assumption, that's why you take the first one in the routing hint instead of the last one.
We can't know for sure whether the ones afterwards are private or not though, but I agree it's a reasonable assumption.

> In the case of mobile wallets hidden behind a hub, getting the id of the mobile wallet is not very useful as we can't contact it directly, I want to know the id of the hub instead.

The id of the mobile wallet wouldn't appear in the routing hints anyway? In a route `A -> B -> mobile-wallet` the routing hints would contain nodeId(A) and nodeId(B). But what you're saying is that B is most likely a private node (otherwise the wallet would have only included `B -> mobile-wallet` in its routing hint).

> I am assuming here that it is extremely rare to have a private node advertising more than one hub in its routing hints, do you think it's unreasonable?

I think we should expect multiple routing hints: Breez and LDK include more than one for example. It doesn't cost us much to store a `Set[PublicKey]` instead of just one, so probably worth doing to avoid another future migration?

So let's keep your assumption about private nodes and log the first `node_id` of every routing hint? Something like `cfg.invoice.flatMap(_.routingInfo.flatMap(_.headOption.map(_.nodeId))).toSet`?

Reply:
I've done it, I expect it to be a rare case but it's definitely cleaner like that.